### ISSUE TITLE
Cancel tasks in Parallel.Invoke

### ISSUE THREAD

**Issue by NikVladi** [dotnet/runtime#63973]
As I understand it, all three static methods in Parallel (For, ForEach and Invoke) create tasks in the background.

You can stop creating these tasks with cancel a token inside ParallelOptions.

I made two simple examples.

In the first uses the For method in the second Invoke.

In the case of the For method, the behavior is expected, after canceling the token, the creation of new tasks is stopped. In the case of the Invoke method, this does not happen. No matter how many methods I put in the Invoke method, they are always executed after the token is canceled. And I don't understand why this is happening.

On [docs.microsoft.com](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.invoke?view=net-6.0) in Parallel.Invoke Method It's been said:

The cancellation token passed in with the ParallelOptions structure enables the caller to cancel the entire operation.

**Question**: Why in the case of the Invoke method, all tasks are executed and canceling the token does nothing? Or maybe I'm doing something wrong, then tell me what.

```
using System;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    class MyClass
    {
        public int a = 0;
        public void Add()
        {
            lock (this)
            {
                Thread.Sleep(100);
                Console.WriteLine($"Do Add {DateTime.Now}");
                a++;
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            void MyMethodForCancel(CancellationTokenSource cancellationTokenSource)
            {
                Random random = new Random();

                while (true)
                {
                    if (random.Next(1, 100) == 50)
                    {
                        cancellationTokenSource.Cancel();
                        Console.WriteLine($"Cancel token {DateTime.Now}");
                        return;
                    }
                }
            }

            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
            ParallelOptions parallelOptions = new ParallelOptions();
            parallelOptions.CancellationToken = cancellationTokenSource.Token;

            MyClass myClass2 = new MyClass();
            Action[] actions = new Action[50];

            for (int i = 0; i < actions.Length; i++)
            {
                actions[i] = myClass2.Add;
            }

            Task MyTask1 = Task.Run(() => Parallel.Invoke(parallelOptions, actions));
            Task MyTask2 = Task.Run(() => { Thread.Sleep(1); MyMethodForCancel(cancellationTokenSource); });

            try
            {
                Task.WaitAll(MyTask1, MyTask2);
            }
            catch
            {

            }

            Console.WriteLine($"a = {myClass2.a}"); //a = 50. Always.               
        }
    }
}
```

**Comment by NikVladi** [comment]
Out of curiosity, I created project under the 4.7.2 framework for the x64 target platform. Solution work correctly as expected.

**Comment by buyaa-n** [comment]
> In the case of the `For` method, the behavior is expected, after canceling the token, the creation of new tasks is stopped. In the case of the Invoke method, this does not happen. No matter how many methods I put in the Invoke method, they are always executed after the token is canceled. And I don't understand why this is happening.

Cancellation token works same for `Invoke` method, i assume its just all tasks already have been created/scheduled when the cancellation requested. I would imagine that for the `Paraller.For` case it would create the tasks sequentially starting from `fromInclusive` to `toExclusive`, therefore you can see that it stops creating new tasks with cancellation.

> Question: Why in the case of the Invoke method, all tasks are executed and canceling the token does nothing? Or maybe I'm doing something wrong, then tell me what.

I suggest to set the [ParallelOptions.MaxDegreeOfParallelism](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.paralleloptions.maxdegreeofparallelism?view=net-6.0#system-threading-tasks-paralleloptions-maxdegreeofparallelism) to some value less than 50 within your code to restrict createing all tasks at once, then you will see that task creation stops with cancellation for `Parallel.Invoke`

**Comment by NikVladi** [comment]
> I suggest to set the [ParallelOptions.MaxDegreeOfParallelism](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.paralleloptions.maxdegreeofparallelism?view=net-6.0#system-threading-tasks-paralleloptions-maxdegreeofparallelism) to some value less than 50 within your code to restrict createing all tasks at once, then you will see that task creation stops with cancellation for `Parallel.Invoke`

You should test your hypothesis by setting the array to 1000 elements. You will see that all 1000 run even on a dual core machine. This method does not stop the initialization of new tasks at all. Setting the MaxDegreeOfParallelism property helps, but the fact remains that if we compare the for, foreach and invoke methods in terms of stopping the initialization of tasks, the invoke method works so that without setting the MaxDegreeOfParallelism property it does not stop the initialization of new tasks at all. In my opinion, this is an obvious bug in the framework.

**Comment by buyaa-n** [comment]
I know it would not stop even for array with 1000 elements, even though for me it makes sense, I don't think it is a bug. Anyway i would defer this questions to the experts in this area CC @stephentoub

**Comment by NikVladi** [comment]
I agree. It would be nice to have an answer:
A) Yes, this is a bug and we will fix it.
b) No, this is not a bug. It's done on purpose. This is done for the following purpose, it gives the following result.
